File: javascript/linkedList/doublyLinkedList.js
class Node{
    constructor(val){
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    push(val){
        let newNode = new Node(val);
        if(!this.head){
            this.head = newNode;
            this.tail = newNode;
        }else{
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }
    pop(){
        if(!this.head) return undefined;
        let currentTail = this.tail
        if(this.length === 1){
            this.head = null;
            this.tail = null;
        }else{
            this.tail = currentTail.prev;
            this.tail.next = null;
            currentTail.prev = null;
        }
        this.length--;
        return currentTail;

    }
    shift(){
        if(this.length === 0) return undefined;
        let oldHead = this.head;
        if(this.length === 1){
            this.head = null;
            this.tail = null;
        }else{
            this.head = oldHead.next;
            this.head.prev = null;
            oldHead.next = null;
        }
        this.length--;
        return oldHead;
    }
    unshift(val){
        let newNode = new Node(val);
        if(this.length === 0){
            this.head = newNode;
            this.tail = newNode;
        }else{
            this.head.prev = newNode;
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return this;
    }
    get(index){
        if(index < 0 || index >= this.length) return null;
        if(index <= (this.length)/2){
            let count = 0;
            var current = this.head;
            while(count != index){
                current = current.next;
                count++;
            }
            return current;
        }else{
            let count = this.length - 1;
            let current = this.tail;
            while(count != index){
                current = current.prev;
                count--
            }return current;
        }
    }
    set(val,index){
        let foundNode = this.get(index);
        if(foundNode != null){
            foundNode.val = val;
            return true
        }
        return false;
    }
    insert(val,index){
        if(index < 0 || index > this.length) return false;
        if(index === 0) return this.unshift(val);
        if(index === this.length) return this.push(val);

        let newNode = new Node(val);
        let BeforeNode=  this.get(index-1);
        let afterNode = this.get(index);

        BeforeNode.next = newNode;
        newNode.prev = BeforeNode;       

        newNode.next = afterNode;
        afterNode.prev = newNode;

        this.length++;
        return true;
    
    }
    remove(index){
        if(index < 0 || index >= this.length) return false;
        if(index === 0) return this.shift();
        if(index === this.length - 1) return this.pop();
        
        let removedNode = this.get(index);
        let beforeNode = removedNode.prev;
        let afterNode = removedNode.next;

        beforeNode.next = afterNode;
        afterNode.prev = beforeNode;
        removedNode.next = null;
        removedNode.prev = null;

        this.length--;
        return removedNode;
    }

}

let list = new DoublyLinkedList();

list.push("786");
list.push("ready");
list.push("to go...")
list.push("hooya")
console.log(list);
console.log(list.remove(2));

console.log(list);

File: python/LinkedList/doublyLinkedList.py
class Node:
    def __init__(self, data=None, prev = None, next=None,):
        self.data = data
        self.prev = prev
        self.next = next

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def isEmpty(self):
        return self.head == None

    def inserttoEnd(self, data):
        node = Node(data)
        if self.isEmpty():
            self.head = node
            self.tail = self.head
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node

    def inserttoStart(self,data):
        node = Node(data)
        if self.isEmpty():
            self.head = node
            self.tail = self.head
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node


    def insertatPosition(self, data, index):

        if index< 0 or index>= self.getLength():
            raise Exception("index out of index")

        if index == 0:
            self.inserttoStart(data)
            return
        count = 0
        itr = self.head
        node = Node(data)
        while itr:
            if count == index - 1:
                itr.next.prev = node
                node.prev = itr
                node.next = itr.next
                itr.next = node
                print(f"{data} is added at index: {index}")
                return
            itr = itr.next 
            count+=1

    def printforward(self):
        itr = self.head
        dlstr = ''
        while itr:
            dlstr += str(itr.data) + '->'
            itr = itr.next
        print(dlstr)

    def printbackward(self):
        itr = self.tail
        dlstr = ''
        while itr:
            dlstr += str(itr.data) + "->"
            itr = itr.prev
        print(dlstr)

    def getLength(self):
        count = 0
        if self.isEmpty():
            return count
        else:
            itr  = self.head
            while itr:
                count+=1
                itr = itr.next
            return count
    
    def deletebyValue(self, val):
        itr = self.head
        while itr.next:
            if itr.next.data == val:
                # print(itr.next.next.data)
                itr.next = itr.next.next
                itr.next.prev = itr 
            itr = itr.next  

if __name__ == "__main__":
    dl = DoublyLinkedList()
    dl.inserttoStart(100)
    dl.inserttoStart(101)
    dl.inserttoEnd(50)
    dl.inserttoEnd(51)
    dl.printforward()
    dl.printbackward()
    dl.insertatPosition(501, 3)
    dl.printforward()
    dl.printbackward()
    dl.inserttoStart(6)
    dl.inserttoEnd(15)
    dl.inserttoEnd(45)
    dl.deletebyValue(15)
    dl.printforward()
    dl.printbackward()



